### DISCUSSION TITLE
How to load data during the viewModel creation??

### DISCUSSION THREAD

**Discussion by michalss** [AvaloniaUI/Avalonia#14330] [General]
Hi,

I have a problem; I cannot load any data into my settings during the model view creation process. Data from the source is loaded, but text fields are not getting populated. Please help.

```csharp
    public partial class SettingsViewModel : ViewModelBase
    {
        public SettingsViewModel()
        {
            LoadData();
        }


        [RelayCommand]
        public async void LoadData()
        {

                 WsUsername = App.Settings.ws_username;
                 OnPropertyChanged(nameof(WsUsername));
                 WsPassword = App.Settings.ws_password;
                 WsToken = App.Settings.ws_token;
      
        }


        [ObservableProperty]
        public string wsUsername = string.Empty;

        [ObservableProperty]
        string wsPassword = string.Empty;

        [ObservableProperty]
        string wsToken = string.Empty;

```

whatever o did try it simple not work.. I did try a few different ways. nothing helped..

```xaml
	<TextBox Grid.Column="1" Watermark="ws username" Text="{Binding WsUsername , Mode=TwoWay}"></TextBox>
```

**Comment by timunie** [comment]
try Dispatcher.UiThread.Post() instead

**Reply by michalss** [reply]
well it does not work, i had to add  OnPropertyChanged to my BaseViewModel class

```csharp
        [RelayCommand]
        public async void LoadData()
        {
            //await Task.Delay(1000);
            //App.Settings.ReadSetting();
            Dispatcher.UIThread.Post(() =>
             {
                 WsUsername = App.Settings.ws_username;
                 OnPropertyChanged(nameof(WsUsername));
                 WsPassword = App.Settings.ws_password;
                 //OnPropertyChanged(nameof(WsPassword));
                 WsToken = App.Settings.ws_token;
                 OnPropertyChanged(nameof(WsToken));
             });
        }
```
this works just fine but why ??


this is very very interesting...

**Reply by timunie** [reply]
How does your ViewModelBase look like?

**Reply by michalss** [reply]
simple it inherriting ObservabelObject and i just added INotifyPropertyChanged..

```csharp

public class ViewModelBase : ObservableObject, INotifyPropertyChanged
{
    public event PropertyChangedEventHandler PropertyChanged;

    protected virtual void OnPropertyChanged(string propertyName)
    {
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
    }
}
```

**Reply by michalss** [reply]
normally i dont need OnPropertyChanged() proc...

**Reply by michalss** [reply]
btw it also works without  Dispatcher.UIThread.Post if i use PropChange. Bug in CommunityToolkit or Avalonia ?

**Reply by timunie** [reply]
I believe something wrong on your side still. Please prepare a very minimal sample for me to double-check.

**Reply by timunie** [reply]
> and i just added INotifyPropertyChanged.

you should not do this as ObservableObject alread implements it for you.

**Comment by michalss** [comment]
> > and i just added INotifyPropertyChanged.
> 
> you should not do this as ObservableObject alread implements it for you.

I know that but it does not work without it. Im pretty sure it si not on my side... but i will try to make a sample later today..

**Reply by michalss** [reply]
finally found the problem... I dont know why but never use combination of small and big letters together with MVVM Community Toolkit.. damn took me a quite a long time to figure it out .. :(

**Reply by damian-666** [reply]
What do you mean by mixing big and small letters, like capitalization? Did you log an issue with the MVVM community, and is there a way to spit out a warning from Avalonia about this if its likely to happen to others?

thisis a suggestion to the contributing .. if anyone else thinks it will help, we all are victims of this kind of thing.   especially with silent fails and nugets..   i  must have lost 6 weeks with monogame on these class of subtle bugs.   

Monitoring hundreds of emails a day isn't feasible,  we all experienced small issues consuming weeks of our lives with are not renewable resources.  .. if its documenting no-one will read or remember it.  it shoulid get spat out at build or runtime..   as a tag, a todo a loud warning.. assert or even expection in rarecases with a mark to revisit when the dependency is fixed.

I'm seeing so much in cross-platform development that can cost hours or weeks unless you're just lucky, and building all the components is a must.

this might be an edge case so mabye its ok to move on fine, but i want to suggest something  general after really watching this churn.

**.
not saying in this case specifically but. when this happens..  
If we all take a little more effort to save others the time cost of these little nuances, we might finish our projects before we die.** The complexity of multiplatform development is unmanageable. I'm only doing Android and Windows, and maybe Mac, but the academic community uses Linux, so I have to do that as well.


i tell myself.. after losing a week on some stupid thing.
-Avoid the compulsion to hack and try stuff. I tend to do that.

-But trace deep into the guts. I eventually might do that; I should do it from the start.


im guilty as well.. For example, in MonoGame, setting the mipmap level to 0 or not 1... something like that... no warning, no fix, OpenGL only. I lost a week or 3 days  .. But I don't know all the low-level graphics code, so I didn't trace it. I found advice on some unrelated issue in some XNA or MonoGame forum on the web,

 and then I should have added a PR that says warns or asserts or spits out a compile time warning.. even if the package is a nuget..   OpenGL this must be 0 , feature is incomplete, or throw exception but that' maybe not be good if its a variable.  . So, warn, don't just draw nothing silently. 
 
 we cant expect people to read all the readmes and docs, or any of them. many  Engineers don't read instructions. For some reason, I'm one of those. I expect things to work as you would expect, to be self-evident.

To save X people 1 week or maybe the product shipping , a bankruptcy, its can be a pile of liltte bugs....

But I just moved on.   many in games do it because they are tired, its over, and forget about other people when we've suffered and in a sense game making is competitive.. but lets face it halving more time to make your game fun is what will determine your talent.    not your hacking perseverance or secret workarounds.   and this is a infrastructure, we are all using it for differing stuff.. 

but things liek 3d, game loops, timers, i have mastered and ill shared it if its relevant, but i dont plan to deploy avalonia, just ot use it for tooling..  not for in game ui .. but im not sure yet.   depends on the weight of skia adn if i can put some ui over my 3d view. now i have some in-game editing .. but that's not ideal.

 i would love it to update at 3000 fps in physics because i do generative design.   im using LLMs to write Ai code for me, they are good at certain homologies.. 

I'm trying to get Microsoft a solution via NLP and the semantic kernel and the new LLMs and the PHI, but visuals will have to work. I'm monitoring over 300 issues a day that are all potential time killers. I can only suggest trying to spare others with any sort of compiler or runtime warnings or something.

Sorry, I haven't even slept. I'm not even committed to Avalonia; I just want it to be stable when I get around to it.  i had to shelf it for my priorities  

 Stride has decided on it. MonoGame is funded, so I don't think I'll use others.   First ones to take over abandoned but undead MSFT things.. should keep on going.

**Reply by timunie** [reply]
@michalss that's why I use `[ObservableProperty] private string _MyString;`. It's nothing related to Avalonia but how the MVVM source generator works.
